src/hooks/utility.rs

```diff
@@ -0,0 +1,152 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#![allow(clippy::too_many_arguments)]
+
+use std::ffi::CString;
+
+use anyhow::{bail, Result};
+use pg_sys::NodeTag;
+use pgrx::*;
+use sqlparser::{ast::Statement, dialect::PostgreSqlDialect, parser::Parser};
+
+use super::query::*;
+
+#[allow(deprecated)]
+type ProcessUtilityHook = fn(
+    pstmt: PgBox<pg_sys::PlannedStmt>,
+    query_string: &core::ffi::CStr,
+    read_only_tree: Option<bool>,
+    context: pg_sys::ProcessUtilityContext::Type,
+    params: PgBox<pg_sys::ParamListInfoData>,
+    query_env: PgBox<pg_sys::QueryEnvironment>,
+    dest: PgBox<pg_sys::DestReceiver>,
+    completion_tag: *mut pg_sys::QueryCompletion,
+) -> HookResult<()>;
+
+pub async fn process_utility_hook(
+    pstmt: PgBox<pg_sys::PlannedStmt>,
+    query_string: &core::ffi::CStr,
+    read_only_tree: Option<bool>,
+    context: pg_sys::ProcessUtilityContext::Type,
+    params: PgBox<pg_sys::ParamListInfoData>,
+    query_env: PgBox<pg_sys::QueryEnvironment>,
+    dest: PgBox<pg_sys::DestReceiver>,
+    completion_tag: *mut pg_sys::QueryCompletion,
+    prev_hook: ProcessUtilityHook,
+) -> Result<()> {
+    let stmt_type = unsafe { (*pstmt.utilityStmt).type_ };
+
+    if !is_support_utility(stmt_type) {
+        prev_hook(
+            pstmt,
+            query_string,
+            read_only_tree,
+            context,
+            params,
+            query_env,
+            dest,
+            completion_tag,
+        );
+
+        return Ok(());
+    }
+
+    let need_exec_prev_hook = match stmt_type {
+        pg_sys::NodeTag::T_ExplainStmt => explain_query(
+            query_string,
+            pstmt.utilityStmt as *mut pg_sys::ExplainStmt,
+            dest.as_ptr(),
+        )?,
+        _ => bail!("unexpected statement type in utility hook"),
+    };
+
+    if need_exec_prev_hook {
+        prev_hook(
+            pstmt,
+            query_string,
+            read_only_tree,
+            context,
+            params,
+            query_env,
+            dest,
+            completion_tag,
+        );
+    }
+
+    Ok(())
+}
+
+fn is_support_utility(stmt_type: NodeTag) -> bool {
+    stmt_type == pg_sys::NodeTag::T_ExplainStmt
+}
+
+fn explain_query(
+    query_string: &core::ffi::CStr,
+    stmt: *mut pg_sys::ExplainStmt,
+    dest: *mut pg_sys::DestReceiver,
+) -> Result<bool> {
+    let query = unsafe { (*stmt).query as *mut pg_sys::Query };
+
+    let query_relations = get_query_relations(unsafe { (*query).rtable });
+    if unsafe { (*query).commandType } != pg_sys::CmdType::CMD_SELECT
+        || !is_duckdb_query(&query_relations)
+    {
+        return Ok(true);
+    }
+
+    if unsafe { !(*stmt).options.is_null() } {
+        error!("No support explain options for DuckDB pushdown query.");
```

Comment:
This error message is a bit confusing. I think you just want to tell the user that their explain option isn't supported, right?

Reply:
The error message has been updated to: 'the EXPLAIN options provided are not supported for DuckDB pushdown queries.'